x/ibc/core/23-commitment/types/merkle.go

```diff
@@ -162,30 +159,37 @@ func (proof MerkleProof) VerifyNonMembership(specs []*ics23.ProofSpec, root expo
 	if !ok {
 		return sdkerrors.Wrapf(ErrInvalidProof, "path %v is not of type MerkleProof", path)
 	}
-	if len(mpath.KeyPath.Keys) != len(specs) {
+	if len(mpath.KeyPath) != len(specs) {
 		return sdkerrors.Wrapf(ErrInvalidProof, "path length %d not same as proof %d",
-			len(mpath.KeyPath.Keys), len(specs))
-	}
-
-	// Convert Proof to []CommitmentProof
-	proofs, err := convertProofs(proof)
-	if err != nil {
-		return err
+			len(mpath.KeyPath), len(specs))
 	}
 
 	// VerifyNonMembership will verify the absence of key in lowest subtree, and then chain inclusion proofs
 	// of all subroots up to final root
-	subroot, err := proofs[0].Calculate()
+	subroot, err := proof.Proofs[0].Calculate()
 	if err != nil {
-		sdkerrors.Wrapf(ErrInvalidProof, "could not calculate root for proof index 0. %v", err)
+		return sdkerrors.Wrapf(ErrInvalidProof, "could not calculate root for proof index 0, merkle tree is likely empty. %v", err)
 	}
-	key := mpath.KeyPath.GetKey(-1)
-	if ok := ics23.VerifyNonMembership(specs[0], subroot, proofs[0], key); !ok {
-		return sdkerrors.Wrapf(ErrInvalidProof, "could not verify absence of key %s", string(key))
+	key, err := mpath.GetKey(uint64(len(mpath.KeyPath) - 1))
+	if err != nil {
+		return sdkerrors.Wrapf(ErrInvalidProof, "could not retrieve key bytes for key: %s", mpath.KeyPath[len(mpath.KeyPath)-1])
+	}
+	switch proof.Proofs[0].Proof.(type) {
+	case *ics23.CommitmentProof_Nonexist:
+		break
```

Comment:
why not just move the code below here instead? All the other conditions return anyways?